python/LLM Examples/PrivateAI_QA_finetune.ipynb

```diff
@@ -0,0 +1,8182 @@
+      "metadata": {
+        "id": "ca151559-7845-46e3-93c6-ffcdfa3f9c79"
+      },
+      "source": [
+        "# Fine-Tuning the Model\n",
+        "In this section, we introduce the fine_tune_model function, designed to streamline the process of fine-tuning our chosen large language model (LLM) on a specific dataset.\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "1a2edb2a-a94e-4e84-bd93-03bb9111c1d6",
+      "metadata": {
+        "tags": [],
+        "id": "1a2edb2a-a94e-4e84-bd93-03bb9111c1d6"
+      },
+      "outputs": [],
+      "source": [
+        "class QADataset(Dataset):\n",
+        "    def __init__(self, data, tokenizer):\n",
+        "        self.data = data\n",
+        "        self.tokenizer = tokenizer\n",
+        "        self.input_ids = []\n",
+        "        self.attn_masks = []\n",
+        "\n",
+        "        max_len_question, max_len_answer = self.find_max_length(data, tokenizer)\n",
+        "        self.max_len = max(max_len_question, max_len_answer)\n",
+        "\n",
+        "        questions = data['question'].tolist()\n",
+        "        answers = data['answer'].tolist()\n",
+        "\n",
+        "        for question, answer in zip(questions, answers):\n",
+        "            encoding = tokenizer(question + tokenizer.eos_token + answer,\n",
+        "                                 truncation=True,\n",
+        "                                 max_length=self.max_len,\n",
+        "                                 padding='max_length',\n",
+        "                                 return_tensors='pt')\n",
+        "            self.input_ids.append(encoding['input_ids'])\n",
+        "            self.attn_masks.append(encoding['attention_mask'])\n",
+        "\n",
+        "    def __len__(self):\n",
+        "        return len(self.input_ids)\n",
+        "\n",
+        "    def __getitem__(self, idx):\n",
+        "        return self.input_ids[idx], self.attn_masks[idx]\n",
+        "\n",
+        "    def find_max_length(self, data, tokenizer):\n",
+        "        max_len_question = 0\n",
+        "        max_len_answer = 0\n",
+        "\n",
+        "        for _, row in data.iterrows():\n",
+        "            tokenized_question = tokenizer.tokenize(row['question'])\n",
+        "            tokenized_answer = tokenizer.tokenize(row['answer'])\n",
+        "\n",
+        "            max_len_question = max(max_len_question, len(tokenized_question))\n",
+        "            max_len_answer = max(max_len_answer, len(tokenized_answer))\n",
+        "\n",
+        "        return max_len_question, max_len_answer\n"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "id": "2e0d5866-2938-4e09-be90-40c86fd88aa3",
+      "metadata": {
+        "tags": [],
+        "id": "2e0d5866-2938-4e09-be90-40c86fd88aa3"
+      },
+      "outputs": [],
+      "source": [
```

Comment:
May want to mention in the docs, this section takes around 30 minutes to complete
![Screenshot 2024-02-27 at 10 20 26 AM](https://github.com/privateai/deid-examples/assets/30555712/c1e8ce4c-4eb8-4e7b-9e44-407142f9c577)